src/diagnostics.ts

```diff
@@ -0,0 +1,62 @@
+/**
+ * this module contains functions to show jest test results in
+ * vscode inspector via the DiagnosticsCollection.
+ */
+import * as vscode from 'vscode'
+// import { DiagnosticCollection, Uri, Diagnostic, Range, DiagnosticSeverity } from 'vscode'
+import { TestFileAssertionStatus } from 'jest-editor-support'
+import { TestReconciliationState } from './TestReconciliationState'
+
+export function updateDiagnositics(testResults: TestFileAssertionStatus[], diagnostics: vscode.DiagnosticCollection) {
+  function addTestFileError(result: TestFileAssertionStatus, uri: vscode.Uri) {
+    const diag = new vscode.Diagnostic(
+      new vscode.Range(0, 0, 0, 0),
+      result.message || 'test file error',
+      vscode.DiagnosticSeverity.Error
+    )
+    diag.source = 'Jest'
+    diagnostics.set(uri, [diag])
+  }
+  function addTestsError(result: TestFileAssertionStatus, uri: vscode.Uri) {
+    const asserts = result.assertions.filter(a => a.status === TestReconciliationState.KnownFail)
+    diagnostics.set(
+      uri,
+      asserts.map(assertion => {
+        const start = 0
+        const daig = new vscode.Diagnostic(
+          new vscode.Range(assertion.line - 1, start, assertion.line - 1, start + 6),
+          assertion.terseMessage || assertion.shortMessage || assertion.message,
+          vscode.DiagnosticSeverity.Error
+        )
+        daig.source = 'Jest'
+        return daig
+      })
+    )
+  }
+
+  testResults.forEach(result => {
+    const uri = vscode.Uri.file(result.file)
+
+    switch (result.status) {
+      case TestReconciliationState.KnownFail:
+        if (result.assertions.length <= 0) {
+          addTestFileError(result, uri)
+        } else {
+          addTestsError(result, uri)
+        }
+        break
+      default:
+        diagnostics.delete(uri)
+        break
+    }
+  })
+}
+
+export function resetDiagnositics(diagnostics: vscode.DiagnosticCollection) {
+  diagnostics.clear()
+}
+export function failedSuiteCount(diagnostics: vscode.DiagnosticCollection): number {
+  let sum = 0
+  diagnostics.forEach(() => sum++)
+  return sum
```

Comment:
interesting, I didn't know `reduce` was not a part of the JS core library

Reply:
Array.prototype.reduce() is indeed part of JS core, unfortunately diagnostics is a `vscode.DiagnosticCollection`, which only exposes forEach() iterator... I know it looked stupid but we have no choice here...

Reply:
👍